src/Elmish.WPF/Logging.fs

```diff
@@ -0,0 +1,130 @@
+    PropertyName: string }
+
+type internal TrySetMemberCalledData =
+  { PropertyNameChain: string
+    PropertyName: string }
+
+type internal GettingErrorsData =
+  { PropertyNameChain: string
+    PropertyName: string option }
+
+
+type internal TraceLogData =
+  | PropertyChangedData of PropertyChangedData
+  | ValidationErrorsChangedData of ValidationErrorsChangedData // TODO: Split by adding ValidationErrorsRemovedData
+  | TimingData of TimingData
+  | CreatingHiddenWindowData of CreatingHiddenWindowData
+  | CreatingVisibleWindowData of CreatingVisibleWindowData
+  | InitializingBindingsData of InitializingBindingsData
+  | ClosingWindowData of ClosingWindowData
+  | HindingWindowData of HindingWindowData
+  | ShowingHiddenWindow of ShowingHiddenWindow
+  | NewSubModelSelectedItemSelectionData of NewSubModelSelectedItemSelectionData
+  | TryGetMemberCalledData of TryGetMemberCalledData
+  | TrySetMemberCalledData of TrySetMemberCalledData
+  | GettingErrorsData of GettingErrorsData
+
+
+let internal logTraceWith logger data =
+  let log fmt = Printf.kprintf logger fmt
+  match data with
+  | PropertyChangedData d -> log "[%s] PropertyChanged \"%s\"" d.PropertyNameChain d.PropertyName
+  | ValidationErrorsChangedData d -> log "[%s] ErrorsChanged \"%s\"" d.PropertyNameChain d.PropertyName
+  | TimingData d -> log "[%s] %s (%ims): %s" d.PropertyNameChain d.BindingDataFunctionName d.ElapsedMilliseconds d.PropertyName
+  | CreatingHiddenWindowData d -> log "[%s] Creating hidden window" d.PropertyNameChain
+  | CreatingVisibleWindowData d -> log "[%s] Creating and opening window" d.PropertyNameChain
+  | InitializingBindingsData d -> log "[%s] Initializing bindings" d.PropertyNameChain
+  | ClosingWindowData d -> log "[%s] Closing window" d.PropertyNameChain
+  | HindingWindowData d -> log "[%s] Hiding window" d.PropertyNameChain
+  | ShowingHiddenWindow d -> log "[%s] Showing existing hidden window" d.PropertyNameChain
+  | NewSubModelSelectedItemSelectionData d -> log "[%s] Setting selected VM to %A" d.PropertyNameChain d.NewSelection
+  | TryGetMemberCalledData d -> log "[%s] TryGetMember %s" d.PropertyNameChain d.PropertyName
+  | TrySetMemberCalledData d -> log "[%s] TrySetMember %s" d.PropertyNameChain d.PropertyName
+  | GettingErrorsData d -> log "[%s] GetErrors %s" d.PropertyNameChain (d.PropertyName |> Option.defaultValue "<null>")
+  
+
+type internal WindowToCloseMissingData =
+  { PropertyNameChain: string }
+
+type internal WindowToHideMissingData =
+  { PropertyNameChain: string }
+
+type internal WindowToShowMissingData =
+  { PropertyNameChain: string }
+
+type internal TryGetMemberMissingBindingData =
+  { PropertyNameChain: string
+    PropertyName: string }
+
+type internal TrySetMemberMissingBindingData =
+  { PropertyNameChain: string
+    PropertyName: string }
+
+type internal TrySetMemberReadOnlyBindingData =
+  { PropertyNameChain: string
+    PropertyName: string }
+
+
+type internal ErrorLogData =
```

Comment:
I don't like us "deciding" on behalf of the user what's errors and what's trace. For example, some of these I might log as warnings, not errors. Perhaps we should have everything in a DU and document each case, explaining what it is sufficiently for the user to decide how to log it?

Reply:
Oh, shoot.  I mean to say
> Serilog has three logging levels _that indicate that something bad happened_

I edited my comment to add that.